core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/worker/DataProcessor.scala

```diff
@@ -329,8 +330,47 @@ class DataProcessor(
           )
           outputIterator.appendSpecialTupleToEnd(FinalizeOperator())
         }
-      case marker: EpochMarker =>
-        epochManager.processEpochMarker(channel.from, marker, this)
+    }
+  }
+
+  def processEpochMarker(
+      channelId: ChannelID,
+      marker: ChannelMarkerPayload,
+      logManager: ReplayLogManager
+  ): Unit = {
+    val markerId = marker.id
+    val command = marker.commandMapping.getOrElse(actorId, null)
+    logger.info(s"receive marker from $channelId, id = ${marker.id}, cmd = ${command}")
+    if (marker.markerType == RequireAlignment) {
+      pauseManager.pauseInputChannel(EpochMarkerPause(markerId), List(channelId))
+    }
+    if (epochManager.isMarkerAligned(upstreamLinkStatus, channelId, marker)) {
+      logManager.markAsReplayDestination(markerId)
+      // invoke the control command carried with the epoch marker
+      logger.info(s"process marker from $channelId, id = ${marker.id}, cmd = ${command}")
+      if (command != null) {
+        asyncRPCServer.receive(command, channelId.from)
+      }
+      // if this operator is not the final destination of the marker, pass it downstream
+      if (!marker.scope.getSinkOperatorIds.contains(getOperatorId)) {
+        val physicalLinks = marker.scope.links.map(_.id).toSet
+        outputManager.flush(Some(physicalLinks))
+        outputGateway.getActiveChannels.foreach { activeChannelId =>
+          if (
+            physicalLinks
+              .exists(p => p.to == VirtualIdentityUtils.getPhysicalOpId(activeChannelId.to))
```

Comment:
can we use ChannelConfig instead to find downstream channels?

Reply:
will change it in another PR.